crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -60,6 +62,16 @@ impl<'db> UnionBuilder<'db> {
         match self.elements.len() {
             0 => Type::Never,
             1 => self.elements[0],
+            2 if self.elements.contains(&Type::BooleanLiteral(true))
```

Comment:
So I should have been clearer in my description of the desired behavior; we want to not just ensure that we can't build a union containing exactly Literal[True] and Literal[False], we want to ensure that any union we build, even if it contains other types also, always replaces those two Literals, if both present, with bool. 

So let's add a test for the case where there is also another type in the union. 

To implement this, I think we'll probably want to add a `simplify` method that gets called before building the final union, similar to what we have for intersections. It can do a single pass over the elements, tracking whether it finds both Boolean literal types, and if so removing them both and adding bool. 

The other way to handle it might be upfront when types are added. This seems like it might be less efficient given the need to search the existing elements for the other literal bool type whenever a literal bool type is added to the union. But this is probably ok in practice, especially if we can do it only when a previously-not-present literal bool type is added to the union. I think whichever seems to work out better in the code can be fine.

Reply:
Whoops that was definitely my bad 😅 Let me know if the current commit is any closer!